deployment/morph/playgrounds.nix

```diff
@@ -0,0 +1,133 @@
+    hostName:
+    let
+      promNodeTextfileDir = pkgs.writeTextDir "roles.prom"
+        ''
+          machine_role{role="playgrounds"} 1
+        '';
+      plutusPlayground = "${plutus-playground.server-invoker}/bin/plutus-playground";
+      marlowePlayground = "${marlowe-playground.server-invoker}/bin/marlowe-playground";
+      playgroundServiceConfig = {
+        TimeoutStartSec = "0";
+        Restart = "always";
+        DynamicUser = true;
+        StateDirectory = [ "pab" ];
+        ProtectKernelTunables = true;
+        ProtectControlGroups = true;
+        ProtectKernelModules = true;
+        PrivateDevices = true;
+        SystemCallArchitectures = "native";
+        ConfigurationDirectory = "playground";
+      };
+      killallz3 = pkgs.writeScriptBin "killallz3" ''
+        kill -9 $(ps aux | grep z3 | grep -v grep | awk '{print $2}')
+      '';
+    in
+    { config, pkgs, lib, ... }:
+    {
+      deployment.secrets = {
+        "plutus-secrets" = {
+          source = "./secrets.plutus.${machines.environment}.env";
+          destination = "/etc/playground/secrets.plutus.${machines.environment}.env";
+          action = [ "systemctl" "restart" "plutus-playground" ];
+        };
+        "marlowe-secrets" = {
+          source = "./secrets.marlowe.${machines.environment}.env";
+          destination = "/etc/playground/secrets.marlowe.${machines.environment}.env";
+          action = [ "systemctl" "restart" "marlowe-playground" ];
+        };
+      };
+      imports = [ (defaultMachine hostName pkgs) ];
+
+      networking.firewall = {
+        enable = true;
+        allowedTCPPorts = with ports; [ ssh plutus-playground-webserver marlowe-playground-webserver nodeExporter ];
+      };
+
+
+      systemd.services.plutus-playground = {
+        enable = true;
+        after = [ "network.target" ];
+        wantedBy = [ "nginx.service" ];
+        before = [ "nginx.service" ];
+        serviceConfig = playgroundServiceConfig // {
+          EnvironmentFile = "/etc/playground/secrets.plutus.${machines.environment}.env";
+        };
+        script = "${plutusPlayground} webserver -p 4000";
+      };
+
+      systemd.services.marlowe-playground = {
+        enable = true;
+        after = [ "network.target" ];
+        wantedBy = [ "nginx.service" ];
+        before = [ "nginx.service" ];
+        serviceConfig = playgroundServiceConfig // {
+          EnvironmentFile = "/etc/playground/secrets.marlowe.${machines.environment}.env";
+        };
+        path = [ pkgs.z3 killallz3 ];
+        script = "${marlowePlayground} webserver -p 4001";
+      };
+
+      services.nginx = {
+        enable = true;
+
+        recommendedGzipSettings = true;
+        recommendedProxySettings = true;
+        recommendedOptimisation = true;
+
+        appendHttpConfig = ''
+          limit_req_zone $binary_remote_addr zone=plutuslimit:10m rate=1r/s;
```

Comment:
what's this for?

Reply:
@krisajenkins added most of these useful nginx defaults a long time ago, I'm leaving them in even though I don't fully understand them all. I've added a small comment.